Adapp 18520 flutter sdk v5 integration guide (#1017)
adjust/dev-docs: src/content/docs/sdk/flutter/features/deep-links.mdx

description: Follow the guides in this section to set up deep linking.
category-title: Deep linking
slug: en/sdk/flutter/features/deep-links
multiVersion: true
---

<SdkVersion version="v5">

You can create deep links to take users to specific pages in your app. The Adjust SDK uses different logic depending on if the user already has your app installed on their device:

-  Direct deep linking: occurs if the user already has your app installed. The link takes the user to the page specified in the link
-  Deferred deep linking: occurs if the user doesn't have your app installed. The link takes the user to a storefront to install your app first. After the user installs the app, it opens to the page specified in the link.

The SDK can read deep link data after a user opens your app from a link.

> ## [Set up deep linking](set-up-deep-linking)

Reviewer: This entire chapter should probably be revised and replaced so that we give Flutter users `tl;dr` from and point them for more info to https://docs.flutter.dev/ui/navigation/deep-linking. That will probably render bunch of stuff from the current chapter as not needed anymore. Let's open a ticket for it to be done later down the road (and ping SDK team to investigate things from there and write the updated chapter). For the time being, I'll comment on the current state of things.

If a user has your app installed, it opens when they interact with a link containing deep link information. The Adjust SDK contains tools to parse deep link information for use throughout your app.

You need to set up deep linking on an app level. Follow the instructions linked below to set up deep linking on your target platform:

-  [iOS](/en/sdk/ios/features/deep-links)
-  [Android](/en/sdk/android/features/deep-links)

## [Reattribution using deep links](reattribution-using-deep-links)

<Callout type="seealso">

Adjust enables you to run re-engagement campaigns with usage of deep links. For more information, see the [guide to appending attribution data to a deep link](https://help.adjust.com/en/article/manage-reattributions#track-reattributions-via-deep-link-clicks).

</Callout>

> To reattribute your user, you need to instantiate an `AdjustDeeplink` object with your deep link URL and pass it to the `Adjust.processDeeplink` method. The Adjust SDK then looks for new attribution data within the deep link. If the SDK finds new information, it forwards the information to Adjust's servers for reattribution.

Reviewer: ```suggestion
To reattribute your user, you need to instantiate an `ADJDeeplink`/`AdjustDeeplink` object with your deep link URL and pass it to the `Adjust.processDeeplink` method. The Adjust SDK then looks for new attribution data within the deep link. If the SDK finds new information, it forwards the information to Adjust's servers for reattribution.
```
Class name depends on the platform / programming language.

<Tabs>
<Tab title="iOS" sync="ios" icon="PlatformIos">

<CodeBlock title="Method signature">

```objc
+ (void)processDeeplink:(ADJDeeplink *)deeplink
```

</CodeBlock>

> To enable deep linking for iOS, call the `appWillOpenUrl` method with a `true` value inside your app delegate.

Reviewer: ```suggestion
To enable Adjust reattribution via deep linking, make sure to call `Adjust.processDeeplink` method once you obtain the deep link in one of your app delegate callbacks.
```

<Tabs>
<Tab title="Swift" sync="swift">

```swift
import Adjust

class AppDelegate {

   func application(application:UIApplication!, didFinishLaunchingWithOptions launchOptions:NSDictionary!) -> Bool {
      GeneratedPluginRegistrant.registerWithRegistry(self)
      // Override point for customization after application launch.
      return super.application(application, didFinishLaunchingWithOptions:launchOptions)
   }

   func application(application:UIApplication!, openURL url:NSURL!, sourceApplication:String!, annotation:AnyObject!) -> Bool {

>       Adjust.appWillOpenUrl(url)

Reviewer: ```suggestion
     let deeplink = ADJDeeplink(deeplink: url)!
     Adjust.processDeeplink(deeplink)
```

      return true
   }

   func application(application:UIApplication!, continueUserActivity userActivity:NSUserActivity!, restorationHandler:([AnyObject]!)->Void) -> Bool {
         if (userActivity.activityType() == NSUserActivityTypeBrowsingWeb) {

>             Adjust.appWillOpenUrl(userActivity.webpageURL())

Reviewer: ```suggestion
            let deeplink = ADJDeeplink(deeplink: userActivity.webpageURL())!
            Adjust.processDeeplink(deeplink)
```

         }
         return true
   }
}
```

</Tab>
<Tab title="Objective-C" sync="objc">

```objc
#import "Adjust.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
   [GeneratedPluginRegistrant registerWithRegistry:self];
   // Override point for customization after application launch.
   return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {

>    [Adjust appWillOpenUrl:url];

Reviewer: ```suggestion
   [Adjust processDeeplink:[ADJDeeplink alloc] initWithDeeplink:url]];
```

   return YES;
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> *restorableObjects))restorationHandler {
   if ([[userActivity activityType] isEqualToString:NSUserActivityTypeBrowsingWeb]) {

>       [Adjust appWillOpenUrl:[userActivity webpageURL]];

Reviewer: ```suggestion
      [Adjust processDeeplink:[ADJDeeplink alloc] initWithDeeplink:[userActivity webpageURL]]];
```

   }
   return YES;
}

@end
```

</Tab>
</Tabs>
</Tab>
<Tab title="Android" sync="android" icon="PlatformAndroid">

<CodeBlock title="Method signature">

```java
public static void processDeeplink(AdjustDeeplink deeplink)
```

</CodeBlock>

> To enable deep linking for Android, call the `appWillOpenUrl` method with a `true` value inside your main activity.

Reviewer: ```suggestion
To enable Adjust reattribution via deep linking, make sure to call `Adjust.processDeeplink` method once you obtain the deep link in your activity.
```

<Tabs>
<Tab title="Kotlin" sync="kotlin">

```kotlin
import com.adjust.sdk.flutter.AdjustSdk
import io.flutter.embedding.android.FlutterActivity
// Used for post flutter 1.12 Android projects
//import io.flutter.app.FlutterActivity

class MainActivity : FlutterActivity() {
   // Either call make the call in onCreate.
   override fun onCreate(savedInstanceState: Bundle?) {
      super.onCreate(savedInstanceState)
      // GeneratedPluginRegistrant.registerWith(this)
      // Used only for pre flutter 1.12 Android projects
      Intent intent = getIntent()
      Uri data = intent.getData()

>       AdjustSdk.appWillOpenUrl(data, this)

Reviewer: ```suggestion
      Adjust.processDeeplink(new AdjustDeeplink(data), this)
```

   }

   // Or make the call in onNewIntent.
   override fun onNewIntent(intent: Intent?) {
      super.onNewIntent(intent)
      Uri data = intent.getData()

>       AdjustSdk.appWillOpenUrl(data, this)

Reviewer: ```suggestion
      Adjust.processDeeplink(new AdjustDeeplink(data), this)
```

   }
}
```

</Tab>
<Tab title="Java" sync="java">

```java
import com.adjust.sdk.flutter.AdjustSdk;
import io.flutter.embedding.android.FlutterActivity; // Used for post flutter 1.12 Android projects
//import io.flutter.app.FlutterActivity; // Used for pre flutter 1.12 Android projects

public class MainActivity extends FlutterActivity {
   // Either call make the call in onCreate.
   @Override
   protected void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);
      // GeneratedPluginRegistrant.registerWith(this); Used only for pre flutter 1.12 Android projects

      Intent intent = getIntent();
      Uri data = intent.getData();

>       AdjustSdk.appWillOpenUrl(data, this);

Reviewer: ```suggestion
      Adjust.processDeeplink(new AdjustDeeplink(data), this);
```

   }

   // Or make the cakll in onNewIntent.
   @Override
   protected void onNewIntent(Intent intent) {
      super.onNewIntent(intent);
      Uri data = intent.getData();

>       AdjustSdk.appWillOpenUrl(data, this);

Reviewer: ```suggestion
      Adjust.processDeeplink(new AdjustDeeplink(data), this);
```

   }
}
```

</Tab>
</Tabs>

</Tab>
</Tabs>

## [Deferred deep linking](deferred-deep-linking)

<CodeBlock title="Property declaration">

```dart
bool? isDeferredDeeplinkOpeningEnabled;
```

</CodeBlock>

The SDK opens deferred deep links by default. No additional setup is required. You can disable this by setting the `isDeferredDeeplinkOpeningEnabled` property on your config instance to **false**.

```dart

> AdjustConfig adjustConfig = new AdjustConfig('{YourAppToken}', AdjustEnvironment.Sandbox);

Reviewer: ```suggestion
AdjustConfig adjustConfig = new AdjustConfig('{YourAppToken}', AdjustEnvironment.sandbox);
```

//...
adjustConfig.isDeferredDeeplinkOpeningEnabled = false;
//...
Adjust.initSdk(adjustConfig);
```

### [Set up a deferred deep link callback](set-up-a-deferred-deep-link-callback)

<CodeBlock title="Property declaration">

```dart

> typedef void DeferredDeeplinkCallback(String? uri);

Reviewer: ```suggestion
typedef void DeferredDeeplinkCallback(String? deeplink);
```
It was correctly documented, yet, we should probably have more meaningful and platform independent name. /cc @MahdiZTD to add the change. ☝️

DeferredDeeplinkCallback? deferredDeeplinkCallback;

```

</CodeBlock>

You can configure the Adjust SDK to call a function when it receives a deferred deep link by adding a function to the `deferredDeeplinkCallback` property on your config instance. This function receives the deep link as a `String` argument.

```dart

> AdjustConfig adjustConfig = new AdjustConfig(yourAppToken, environment);

Reviewer: ```suggestion
AdjustConfig adjustConfig = new AdjustConfig('{YourAppToken}', AdjustEnvironment.sandbox);
```

> adjustConfig.deferredDeeplinkCallback = (String uri) {

Reviewer: ```suggestion
adjustConfig.deferredDeeplinkCallback = (String? deeplink) {
```
/cc @MahdiZTD ☝️

>    print('[Adjust]: Received deferred deeplink: ' + uri);

Reviewer: ```suggestion
   print('[Adjust]: Received deferred deeplink: ' + deeplink);
```
/cc @MahdiZTD ☝️

};
Adjust.initSdk(adjustConfig);
```

> ### [Example](example)

Reviewer: Code snippet from above is explaining the same thing as this chapter. We can probably remove the chapter.

   config.logLevel = AdjustLogLevel.verbose;

   config.deferredDeeplinkCallback = (String? uri) {
      print('[Adjust]: Received deferred deeplink: ' + uri!);
   };

   Adjust.initSdk(config);
}
```

### [Enable LinkMe](enable-linkme)

<CodeBlock title="Property declaration">

```dart
bool? isLinkMeEnabled;
```

</CodeBlock>

The Adjust SDK lets you copy deep link information from the device pasteboard. When combined with Adjust’s LinkMe solution, this feature enables deferred deep linking on devices running iOS 15 and above.

<Callout type="important">

The Adjust SDK checks the pasteboard when a user opens the app for the first time. The device displays a dialog asking if the user wants to allow the app to read the pasteboard.

</Callout>

When a user clicks on a LinkMe URL they have the option to copy the link information to their system pasteboard. You can use the Adjust SDK to read the system pasteboard for deep link information. If deep link information is present, the SDK forwards the user to the correct page in your app.

To enable pasteboard checking in your app, set the `isLinkMeEnabled` property to `true` on your config object:

```dart

> AdjustConfig adjustConfig = new AdjustConfig('{YourAppToken}', AdjustEnvironment.Sandbox);

Reviewer: ```suggestion
AdjustConfig adjustConfig = new AdjustConfig('{YourAppToken}', AdjustEnvironment.sandbox);
```
